Fixed error for suggestions not available, also fixed errors for missing rendered objects

nik1097 opened the pull request:
Fixed "Key Error" occurring for queries that don't have suggestions using a try catch block.
Also fixed issues #1202, #1189, #1187, #1162
Handled the following rendered objects errors by adding "if statements":
1. hashtagTileRenderer
2. includingResultsForRenderer
3. showingResultsForRenderer
4. movieRenderer

Could you please review the Pull Request and confirm merger? @nficano @hbmartin @tfdahlin @josegonzalez 
Thanks in advance!

nik1097 committed:
fixed error when suggestions aren't available and also fixed errors for rendered objects

UnknownCrafts commented:
why hasn't this been merged yet? I opened a new pull request for no reason

nik1097 commented:
@UnknownCrafts thanks! I've been trying to notify the owners for some time now. Hoping this gets merged soon :/

UnknownCrafts reviewed (approved):
These changes look good

UnknownCrafts commented:
and now we wait for people with write access

nik1097 commented:
@tfdahlin could you please review this change and merge it? It's a simple change. I can see you have write access on this repo. Thank you so much!

nik1097 commented:
@nficano @RONNCC Heyy could you please review the changes and merge this with the master? I need this is a requirement for my college grade, thank you so much! It's a small change and I've documented all of the changes with comments.

josegonzalez commented:
It looks like I don't have commit bit here anymore, but maybe that's the GitHub mobile app ui?

nik1097 commented:
@josegonzalez yeah could you please let me know if you could check it through your computer? Thank you so much!

josegonzalez commented:
I checked and I don't think I have access.

Do you need a release or just a way to install? You can install an editable version of your branch if the latter.

nik1097 commented:
I just want my pull request to be reviewed and merged with the Master if the changes I made make sense?

nik1097 reviewed (commented):
Could you please merge these changes to the master? @UnknownCrafts

UnknownCrafts commented:
> Could you please merge these changes to the master? @UnknownCrafts

sorry but I don't have write access

RONNCC commented:
hey @nik1097  

Thanks for the PR!

If this is still relevant, could you add the following:

- a unit test to show the behavior before and after
- some context on the need for the change

Ideally this would 'break' in the future when youtube also changes behavior again and we can show it as a regression.
In addition it would help document the change in the code using tests.

nik1097 commented:
Hey @RONNCC 

Thank you so much for your response and apologies for the delay in my response. 
Please find the context below.
Program breaks with a Key Error when no suggestions are returned while querying for word suggestions. I've added a try catch block to fix this error.
Users also report missing rendered objects error when they query for words that are unavailable, I've added if statements similar to ones that exist that handle other such objects.

Also, as I'm new to this. Do you need the unit test as a file in the repository or can I attach images for the unit test I have carried out locally?

Regards,
Nikhil

nik1097 committed:
added unit test file to verify changes made

nik1097 commented:
Hey @RONNCC I have added unit test file to verify the changes that I made, could you please go through it and suggest any changes if required? Thanks in advance!

glubsy commented on tests/test_suggestions.py:
```diff
@@ -0,0 +1,16 @@
+    s = Search(temp_2)
+    assert s.completion_suggestions
+
+test_completion_suggestions()
```
Do not call the function yourself. Read the basics of [pytest documentation](https://docs.pytest.org/en/latest/getting-started.html) for more information.

RONNCC replied:
+1

also assert "s.completion_suggestions" is asserting the boolean quantity of the object which is a bit difficult to parse at first read -- generally easy to assert notEmpty or similar

nik1097 replied:
I'm currently working on changing this, thank you for your feedback @RONNCC I will get back to you asap.

nik1097 replied:
I've moved the test file to the right path as suggested by @glubsy and I will change the code that conforms to the testing standards and push a commit soon.

nik1097 replied:
@glubsy I'm trying to implement the code using mocks as you suggested, I've stored 3 different .html files (for the cases I've implemented the fix) in the mock folder. I'm trying to directly access these .html files using my test code for simulating the requests. Could you please provide your advice on this and confirm if I'm heading in the right direction? After completing my testing on the mock folder, do I zip it? Thanks in advance!

glubsy replied:
Let's keep this discussion in the main thread instead.

glubsy commented:
Make sure your changes conform to pep8 by running flake8 on it.

nik1097 commented:
@glubsy thanks for pointing me in the right direction, I will make the changes!

glubsy commented:
Move the test file into `tests/contrib` and rename it `test_search.py`. 
Keep the style consistent with the other test files. 
And try to avoid making requests if possible: save the results in the `tests/mocks` directory, and use mocks and fixtures to simulate requests without actually going over the wire.

nik1097 committed:
updated file to conform to pep8

nik1097 commented:
> Make sure your changes conform to pep8 by running flake8 on it.

Made the necessary changes to the file to conform to pep8.

RONNCC commented on pytube/contrib/search.py:
```diff
@@ -42,8 +42,12 @@ def completion_suggestions(self):
-            self._completion_suggestions = self._initial_results['refinements']
-        return self._completion_suggestions
+            try:
+                if self._initial_results['refinements']:
```
why is there an if here?

nik1097 replied:
Hey thanks for pointing this out, this was redundant and hence I've gotten rid of it.

nik1097 committed:
removed redundant if statement & new test dir

glubsy commented:
> I'm trying to implement the code using mocks as you suggested, I've stored 3 different .html files (for the cases I've implemented the fix) in the mock folder. I'm trying to directly access these .html files using my test code for simulating the requests. Could you please provide your advice on this and confirm if I'm heading in the right direction? After completing my testing on the mock folder, do I zip it? Thanks in advance!

Looks like you're heading in the right direction yes. Look at the other test files for examples, you'll figure it out by studying how it is implemented (look at confest.py). Use gzip.

nik1097 committed:
updated the test file with better error checks

nik1097 commented:
@RONNCC I have updated the test function, I think now the assert statement makes more sense. Please let me know your thoughts. As pointed out by @glubsy I'm still working my way around creating a mock function instead of directly accessing the API. I've seen quite a few test files that have directly accessed the API in the repo. 
I might need a little more time to implement the mock feature for testing. But I wanted to know if my current fix will suffice? Could you please review the changes and let me know if it makes sense? Thank you so much for your continued support!

nik1097 commented:
Could you please review the latest changes I've made? @RONNCC

glubsy commented on tests/contrib/test_search.py:
```diff
@@ -0,0 +1,36 @@
+        if 'Unexpected renderer encountered.' not in s.completion_suggestions:
+            assert True
+    except KeyError:
+        assert False
```
Maybe it's just me, but I would suggest rewriting these tests. 
They are non-deterministic since they fetch data over the wire, that's why using mocks (and fixtures) is important here.

nik1097 replied:
Hey, @glubsy thanks for your input. Alright, I'll try spending some time over the weekend to get this done. I'm having a hard time trying to create mocks for this use case, but I will try again. Thanks!

nik1097 committed:
completed unittest with unittesting module

nik1097 committed:
used mocks for creating test cases for search

nik1097 commented:
Hey @glubsy I apologize for the delay. I was having some personal commitments. I have rewritten my test cases with mocks. Could you please review it? I also wrote it using unit tests in my previous commit. I understand you didn't want me to use the API calls for my tests. Please suggest any changes that you would like me to incorporate. Thanks for your time.

nik1097 commented:
Hey @RONNCC @nficano could you please review my changes? I've added the unit test file as requested using mocks.

glubsy commented:
Tests fail on my side, the "search_suggestion" fixture is missing.

RONNCC commented:
@glubsy  can you help ensure this works + review it -- thanks

glubsy commented:
```log
> pytest -k test_search 
================================================================================================ test session starts ================================================================================================
platform linux -- Python 3.10.4, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
rootdir: /pytube
collected 216 items / 212 deselected / 4 selected                                                                                                                                                                   

tests/contrib/test_search.py EEEE                                                                                                                                                                             [100%]

====================================================================================================== ERRORS =======================================================================================================
______________________________________________________________________________ ERROR at setup of test_for_missing_hashtagTileRenderer _______________________________________________________________________________
file /pytube/tests/contrib/test_search.py, line 4
  @mock.patch('pytube.request.get')
  def test_for_missing_hashtagTileRenderer(request_get, search_suggestion):
E       fixture 'search_suggestion' not found
>       available fixtures: age_restricted, cache, capfd, capfdbinary, caplog, capsys, capsysbinary, channel_videos_html, cipher_signature, doctest_namespace, missing_recording, monkeypatch, playlist_html, playlist_long_html, playlist_submenu_html, presigned_video, private, pytestconfig, record_property, record_testsuite_property, record_xml_attribute, recwarn, stream_dict, tmp_path, tmp_path_factory, tmpdir, tmpdir_factory
>       use 'pytest --fixtures [testpath]' for help on them.

/pytube/tests/contrib/test_search.py:4
__________________________________________________________________________________ ERROR at setup of test_test_for_no_suggestions ___________________________________________________________________________________
file /pytube/tests/contrib/test_search.py, line 10
  @mock.patch('pytube.request.get')
  def test_test_for_no_suggestions(request_get, search_suggestion):
E       fixture 'search_suggestion' not found
>       available fixtures: age_restricted, cache, capfd, capfdbinary, caplog, capsys, capsysbinary, channel_videos_html, cipher_signature, doctest_namespace, missing_recording, monkeypatch, playlist_html, playlist_long_html, playlist_submenu_html, presigned_video, private, pytestconfig, record_property, record_testsuite_property, record_xml_attribute, recwarn, stream_dict, tmp_path, tmp_path_factory, tmpdir, tmpdir_factory
>       use 'pytest --fixtures [testpath]' for help on them.

/pytube/tests/contrib/test_search.py:10
_________________________________________________________________________________ ERROR at setup of test_for_missing_movieRenderer __________________________________________________________________________________
file /pytube/tests/contrib/test_search.py, line 16
  @mock.patch('pytube.request.get')
  def test_for_missing_movieRenderer(request_get, search_suggestion):
E       fixture 'search_suggestion' not found
>       available fixtures: age_restricted, cache, capfd, capfdbinary, caplog, capsys, capsysbinary, channel_videos_html, cipher_signature, doctest_namespace, missing_recording, monkeypatch, playlist_html, playlist_long_html, playlist_submenu_html, presigned_video, private, pytestconfig, record_property, record_testsuite_property, record_xml_attribute, recwarn, stream_dict, tmp_path, tmp_path_factory, tmpdir, tmpdir_factory
>       use 'pytest --fixtures [testpath]' for help on them.

/pytube/tests/contrib/test_search.py:16
_______________________________________________________________________________ ERROR at setup of test_for_showingResultsForRenderer ________________________________________________________________________________
file /pytube/tests/contrib/test_search.py, line 22
  @mock.patch('pytube.request.get')
  def test_for_showingResultsForRenderer(request_get, search_suggestion):
E       fixture 'search_suggestion' not found
>       available fixtures: age_restricted, cache, capfd, capfdbinary, caplog, capsys, capsysbinary, channel_videos_html, cipher_signature, doctest_namespace, missing_recording, monkeypatch, playlist_html, playlist_long_html, playlist_submenu_html, presigned_video, private, pytestconfig, record_property, record_testsuite_property, record_xml_attribute, recwarn, stream_dict, tmp_path, tmp_path_factory, tmpdir, tmpdir_factory
>       use 'pytest --fixtures [testpath]' for help on them.

/pytube/tests/contrib/test_search.py:22
============================================================================================== short test summary info ==============================================================================================
ERROR tests/contrib/test_search.py::test_for_missing_hashtagTileRenderer
ERROR tests/contrib/test_search.py::test_test_for_no_suggestions
ERROR tests/contrib/test_search.py::test_for_missing_movieRenderer
ERROR tests/contrib/test_search.py::test_for_showingResultsForRenderer
========================================================================================= 212 deselected, 4 errors in 0.13s =========================================================================================
```

@nik1097 looks like it's missing the file where the "search_suggestion" fixture is defined.

RONNCC commented:
not sure if this is still relevant .... (?)

RONNCC commented:
@glubsy  any ideas?
